File: compile.js
#!/usr/bin/env node
// compile.js — Fetches all 花间集 volumes and compiles into poems.json
const https = require('https');
const fs = require('fs');

const BASE = 'https://raw.githubusercontent.com/chinese-poetry/chinese-poetry/master/五代诗词/huajianji';
const VOLUMES = [
  'huajianji-1-juan.json',
  'huajianji-2-juan.json',
  'huajianji-3-juan.json',
  'huajianji-4-juan.json',
  'huajianji-5-juan.json',
  'huajianji-6-juan.json',
  'huajianji-7-juan.json',
  'huajianji-8-juan.json',
  'huajianji-9-juan.json',
  'huajianji-x-juan.json',
];

function fetchJSON(url) {
  return new Promise((resolve, reject) => {
    https.get(url, (res) => {
      let data = '';
      res.on('data', (chunk) => data += chunk);
      res.on('end', () => {
        try { resolve(JSON.parse(data)); }
        catch (e) { reject(e); }
      });
    }).on('error', reject);
  });
}

async function main() {
  let allPoems = [];
  let id = 1;
  const authors = new Set();

  for (const vol of VOLUMES) {
    const url = `${BASE}/${vol}`;
    console.log(`Fetching ${url}...`);
    const poems = await fetchJSON(url);
    for (const p of poems) {
      authors.add(p.author);
      // Split author if multiple (e.g., "尹鹗 六首" -> "尹鹗")
      const cleanAuthor = p.author.replace(/[。\s]*\d+[首\s]*$/, '').trim();
      allPoems.push({
        id: id++,
        title: p.title || '',
        author: cleanAuthor || p.author,
        rhythmic: p.rhythmic || '',
        content: p.paragraphs || [],
        period: '五代',
        tags: [],
      });
    }
  }

  // Simple tag classification based on keywords in title and content
  const keywordTags = {
    '春': 'spring',
    '秋': 'autumn',
    '花': 'flower',
    '月': 'moon',
    '柳': 'willow',
    '雨': 'rain',
    '雪': 'snow',
    '莲': 'lotus',
    '梅': 'plum',
    '梦': 'dream',
    '愁': 'sorrow',
    '恨': 'regret',
    '泪': 'tears',
    '相思': 'longing',
    '离别': 'parting',
    '忆': 'memory',
    '归': 'return',
    '闺': 'boudoir',
    '妆': 'adornment',
  };

  for (const poem of allPoems) {
    const text = (poem.title + poem.content.join('')).toLowerCase();
    for (const [kw, tag] of Object.entries(keywordTags)) {
      if (text.includes(kw) && !poem.tags.includes(tag)) {
        poem.tags.push(tag);
      }
    }
  }

  const output = { meta: { title: '花间集', period: '五代', total: allPoems.length, authors: [...authors].sort() }, poems: allPoems };
  fs.writeFileSync('poems.json', JSON.stringify(output, null, 2), 'utf8');
  console.log(`Done! Compiled ${allPoems.length} poems by ${authors.size} authors → poems.json`);
}

main().catch(console.error);

File: compile.py
#!/usr/bin/env python3
"""compile.py — Fetches all 花间集 volumes and compiles into poems.json"""
import json, urllib.request, sys, re

# Use already-fetched data — write poems.json from inline data
# Since we can't fetch with Unicode URLs, we'll create the JSON programmatically
# from the raw data we already downloaded.

# Read each volume from the local copies (if available) or construct manually
# We'll use the data stored in the tool's context

VOLUME_URLS = [
    "https://raw.githubusercontent.com/chinese-poetry/chinese-poetry/master/%E4%BA%94%E4%BB%A3%E8%AF%97%E8%AF%8D/huajianji/huajianji-1-juan.json",
    "https://raw.githubusercontent.com/chinese-poetry/chinese-poetry/master/%E4%BA%94%E4%BB%A3%E8%AF%97%E8%AF%8D/huajianji/huajianji-2-juan.json",
    "https://raw.githubusercontent.com/chinese-poetry/chinese-poetry/master/%E4%BA%94%E4%BB%A3%E8%AF%97%E8%AF%8D/huajianji/huajianji-3-juan.json",
    "https://raw.githubusercontent.com/chinese-poetry/chinese-poetry/master/%E4%BA%94%E4%BB%A3%E8%AF%97%E8%AF%8D/huajianji/huajianji-4-juan.json",
    "https://raw.githubusercontent.com/chinese-poetry/chinese-poetry/master/%E4%BA%94%E4%BB%A3%E8%AF%97%E8%AF%8D/huajianji/huajianji-5-juan.json",
    "https://raw.githubusercontent.com/chinese-poetry/chinese-poetry/master/%E4%BA%94%E4%BB%A3%E8%AF%97%E8%AF%8D/huajianji/huajianji-6-juan.json",
    "https://raw.githubusercontent.com/chinese-poetry/chinese-poetry/master/%E4%BA%94%E4%BB%A3%E8%AF%97%E8%AF%8D/huajianji/huajianji-7-juan.json",
    "https://raw.githubusercontent.com/chinese-poetry/chinese-poetry/master/%E4%BA%94%E4%BB%A3%E8%AF%97%E8%AF%8D/huajianji/huajianji-8-juan.json",
    "https://raw.githubusercontent.com/chinese-poetry/chinese-poetry/master/%E4%BA%94%E4%BB%A3%E8%AF%97%E8%AF%8D/huajianji/huajianji-9-juan.json",
    "https://raw.githubusercontent.com/chinese-poetry/chinese-poetry/master/%E4%BA%94%E4%BB%A3%E8%AF%97%E8%AF%8D/huajianji/huajianji-x-juan.json",
]

def fetch_json(url):
    req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(req) as resp:
        return json.loads(resp.read().decode("utf-8"))

def main():
    all_poems = []
    authors = set()
    pid = 1

    for url in VOLUME_URLS:
        print(f"Fetching {url}...", file=sys.stderr)
        poems = fetch_json(url)
        for p in poems:
            author = p.get("author", "").strip()
            clean_author = re.sub(r'[。\s]*\d+[首\s]*$', '', author).strip() or author
            authors.add(clean_author)
            all_poems.append({
                "id": pid,
                "title": p.get("title", ""),
                "author": clean_author,
                "rhythmic": p.get("rhythmic", ""),
                "content": p.get("paragraphs", []),
                "period": "五代",
                "tags": [],
            })
            pid += 1

    # Keyword tagging
    keyword_tags = {
        "春": "spring", "秋": "autumn", "花": "flower", "月": "moon",
        "柳": "willow", "雨": "rain", "雪": "snow", "莲": "lotus",
        "梅": "plum", "梦": "dream", "愁": "sorrow", "恨": "regret",
        "泪": "tears", "相思": "longing", "离别": "parting", "忆": "memory",
        "归": "return", "闺": "boudoir", "妆": "adornment",
    }
    for poem in all_poems:
        text = poem["title"] + "".join(poem["content"])
        for kw, tag in keyword_tags.items():
            if kw in text and tag not in poem["tags"]:
                poem["tags"].append(tag)

    output = {
        "meta": {
            "title": "花间集",
            "period": "五代",
            "total": len(all_poems),
            "authors": sorted(authors),
        },
        "poems": all_poems,
    }
    with open("poems.json", "w", encoding="utf-8") as f:
        json.dump(output, f, ensure_ascii=False, indent=2)
    print(f"Done! {len(all_poems)} poems, {len(authors)} authors → poems.json")

if __name__ == "__main__":
    main()
